1-theory/1-apis-services-microservices.adoc

```diff
@@ -0,0 +1,144 @@
+API, whilst a Java library would also consider the method names and arguments to be
+an API.
+
+Essentially it’s all the same concept, but when talking about Web APIs we’re talking
+about the utilization of network protocols such as HTTP, AMQP, etc., and instead of
+command-line arguments or method parameters, we’re working with URLs, query string
+parameters, or chunks of data (often JSON).
+
+The goal of a Web API is to provide other applications with access to specific subset of
+functionality and data that this application owns. Some APIs are public (anyone can get
+the weather from api.weather.gov), some are private (it’s probably tough to get access
+to JPMorgan Chase’s Enterprise Customers API), and some are a mixture.
+
+APIs can give you data for a single-page JavaScript application, handle payment
+information so you don’t have to worry about storing credit card details, post a
+Facebook status on a users timeline, or share the same data across a myriad of different
+devices. Watch that Netflix show on your Xbox, it’s all good, APIs have you covered.
+
+APIs are built by all sorts of folks; maybe another company, some large like Google or
+Facebook, startups, governments, or charity organizations. They could also be
+coworkers on another team, or another department, on another floor (or continent).
+
+== Service
+
+A service is another very overloaded term in computer science, and programming in genral. A lot 
+of developers are used to MVC (Model, View, Controller) and a service is just "any code that didnt
+fit into a model, view, or controller". 
+
+When folks talk about services in API-land, they are usually talking about "Service-orientated Architecture".
+
+[quote,Wikipedia,https://wikipedia.org/wiki/Service-oriented_architecture]
+____
+Service-oriented architecture (SOA) is a style of software design where services are provided to the other components by application components, through a communication protocol over a network. The basic principles of service-oriented architecture are independent of vendors, products and technologies. A service is a discrete unit of functionality that can be accessed remotely and acted upon and updated independently, such as retrieving a credit card statement online.
+
+A service has four properties according to one of many definitions of SOA:
+
+- It logically represents a business activity with a specified outcome.
+- It is self-contained.
+- It is a black box for its consumers.
+- It may consist of other underlying services.
+
+SOA was first termed Service-Based Architecture in 1998 by a team developing integrated foundational management services and then business process-type services based upon units of work and using CORBA for inter-process communications. 
+____
+
+Service-orientated architecture is essentially the idea of taking various modules out of your application, and putting them on a different server, 
```

Comment:
I think my confusion in

https://twitter.com/ramimassoud/status/1103127686103396353

is based on my misunderstanding of the definition of SOA in [SOA Patterns](https://www.amazon.com/gp/product/1933988266/)

> _Service-oriented architecture_ (SOA) is an architectural style for building systems based on interactions of loosely coupled, coarse-grained, and autonomous components called _services_. Each service exposes processes and behavior through _contracts_, which are composed of _messages_ at discoverable addresses called _endpoints_. A service's behavior is governed by policies that are external to the service itself. The contracts and messages are used by external components called _service consumers_.

I haven't been able to dig into the book yet, so that's likely most of my issue. You're right when you said

> Even if they live on the same server they’re using the local network interface so that’s a technicality.

and I think even if two services communicated over a Unix socket, that's the same deal. I was wondering if you could still call it SOA if you did (what I think Shopify does?) and broke down functionality in a Rails app into Rails Engines and Gems? At least as long as the different components didn't rely on global shared access to the DB and they had to communicate with each other through defined APIs/methods.

With all that said, maybe ditching talking about servers would clarify a bit?

```suggestion
Service-oriented architecture is essentially the idea of breaking various pieces of functionality out of your application and into their own autonomous applications
```

Reply:
I've loosened the paragraph a bit based on this feedback:

> Service-oriented architecture is essentially the idea of breaking various
modules out of your application, possibly putting them on different servers, with a
network interface or socket facilitating communication.

I dunno what Shopify are doing but based on your description it sounds like theyre just sharing some reusable code via gems, thats not really SoA in this sense. You could call it that, but you can call anything anything. :)